equinox/nn/_embedding.py

```diff
@@ -89,3 +96,313 @@ def __call__(
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(seq_len, embedding_size)`, with the rotary positional
+        encoding applied to the input.
+        """
+
+        seq_len, embedding_size = x.shape
+        if embedding_size != self.embedding_size:
+            raise ValueError(
+                f"x.shape[-1] must match self.embedding_size, "
+                f"but {x.shape[-1]} != {self.embedding_size}"
+            )
+        if embedding_size % 2 != 0:
+            raise ValueError(
+                f"x.shape[-1] must be even, but {x.shape[-1]} is not even."
+            )
+
+        neg_half_x = self.negate_half(x)
+
+        with jax.ensure_compile_time_eval():
+            if (embedding_size, seq_len) in internal_rope_embedding_cache:
+                freqs_cis = internal_rope_embedding_cache[(embedding_size, seq_len)]
+            else:
+                freqs_cis = self.precompute_freqs_cis(embedding_size, seq_len)
+                internal_rope_embedding_cache[(embedding_size, seq_len)] = freqs_cis
+
+        assert freqs_cis is not None, "freqs_cis must not be None."
+        freqs_real = jnp.tile(freqs_cis.real, (1, 2))
+        freqs_imag = jnp.tile(freqs_cis.imag, (1, 2))
+
+        x_rope = (x * freqs_real) + (neg_half_x * freqs_imag)
+        return jax.lax.stop_gradient(x_rope)
+
+
+class SinusoidalPositionalEmbedding(Module):
+    r"""
+    A sinusoidal positional encoding module, as described in the paper
+    "Attention is All You Need". While this module can be used in any context, it is
+    particularly useful for providing positional information to transformer models.
+
+    !!! example
+        The following example demonstrates how to use `SinusoidalPositionalEmbedding` in
+        a simple transformer model.
+        ```python
+
+        class TransformerBlock(eqx.nn.StatefulLayer):
+            ...
+            key_embeddings: SinusoidalPositionalEmbedding
+            query_embeddings: SinusoidalPositionalEmbedding
+
+            def __init__(...):
+                ...
+                self.query_embeddings = SinusoidalPositionalEmbedding(
+                    embedding_size=n_embd
+                )
+                self.key_embeddings = SinusoidalPositionalEmbedding(
+                    embedding_size=n_embd
+                )
+                ...
+
+            def __call__(...):
+                def process_heads(query_heads, key_heads, value_heads):
+                    query_heads = jax.vmap(self.query_embeddings,
+                                           in_axes=1,
+                                           out_axes=1)(query_heads)
+                    key_heads = jax.vmap(self.key_embeddings,
+                                         in_axes=1,
+                                         out_axes=1)(key_heads)
+
+                    return query_heads, key_heads, value_heads
```

Comment:
Possibly I'm revealing some naivety here, but I don't think I've seen sinusoidal embeddings used to modify embeddings very frequently. (Ever?) I usually think of it as something being provided at the very start of the architecture, not something provided on each layer.
@dlwh -- you probably have a better idea of this than I do?

Reply:
Oh, I think my brain went afk when I wrote this. You're absolutely right, sinusoidal embeddings happen before the query/key/value projections, making this just plain wrong. Don't know what happened there. I'll come up with a better example. 😄